spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/autoconfig/TraceAutoConfiguration.java

```diff
@@ -143,50 +146,53 @@ SpanNamer sleuthSpanNamer() {
 		return new DefaultSpanNamer();
 	}
 
+	/**
+	 * To override the underlying context format, override this bean and set the delegate
+	 * to what you need. {@link BaggagePropagation.FactoryBuilder} will unwrap itself if
+	 * no fields are configured.
+	 */
 	@Bean
 	@ConditionalOnMissingBean
-	Propagation.Factory sleuthPropagation(SleuthProperties sleuthProperties) {
-		if (sleuthProperties.getBaggageKeys().isEmpty()
-				&& sleuthProperties.getPropagationKeys().isEmpty()
-				&& extraFieldCustomizers.isEmpty()
-				&& this.extraFieldPropagationFactoryBuilder == null
-				&& sleuthProperties.getLocalKeys().isEmpty()) {
-			return B3Propagation.FACTORY;
-		}
-		ExtraFieldPropagation.FactoryBuilder factoryBuilder;
-		if (this.extraFieldPropagationFactoryBuilder != null) {
-			factoryBuilder = this.extraFieldPropagationFactoryBuilder;
-		}
-		else {
-			factoryBuilder = ExtraFieldPropagation
-					.newFactoryBuilder(B3Propagation.FACTORY);
-		}
-		if (!sleuthProperties.getBaggageKeys().isEmpty()) {
-			factoryBuilder = factoryBuilder
-					// for HTTP
-					.addPrefixedFields("baggage-", sleuthProperties.getBaggageKeys())
-					// for messaging
-					.addPrefixedFields("baggage_", sleuthProperties.getBaggageKeys());
-		}
-		if (!sleuthProperties.getPropagationKeys().isEmpty()) {
-			for (String key : sleuthProperties.getPropagationKeys()) {
-				factoryBuilder = factoryBuilder.addField(key);
+	BaggagePropagation.FactoryBuilder baggagePropagationFactoryBuilder() {
+		return BaggagePropagation.newFactoryBuilder(B3Propagation.FACTORY);
+	}
+
+	@Bean
+	@ConditionalOnMissingBean
+	Propagation.Factory sleuthPropagation(
+			BaggagePropagation.FactoryBuilder factoryBuilder, List<BaggageField> baggage,
+			SleuthProperties sleuthProperties) {
+
+		Set<String> local = new TreeSet<>(String.CASE_INSENSITIVE_ORDER);
+		local.addAll(sleuthProperties.getLocalKeys());
+
+		Set<String> toPrefix = new TreeSet<>(String.CASE_INSENSITIVE_ORDER);
+		toPrefix.addAll(sleuthProperties.getBaggageKeys());
+
+		for (BaggageField field : baggage) {
+			if (local.contains(field.name())) {
+				factoryBuilder.addField(field);
 			}
-		}
-		if (!sleuthProperties.getLocalKeys().isEmpty()) {
-			for (String key : sleuthProperties.getLocalKeys()) {
-				factoryBuilder = factoryBuilder.addRedactedField(key);
+			else if (toPrefix.contains(field.name())) {
+				List<String> prefixedNames = new ArrayList<>();
+				prefixedNames.add("baggage-" + field.name()); // for HTTP
```

Comment:
ps this is super chatty, to encode both ways. we should stop doing this at some point. ideally we can move onto the baggage header format being discussed in w3c (used to be named correlation-context)